src/prng/chacha.rs

```diff
@@ -129,66 +121,53 @@ impl ChaChaRng {
-    /// counter of ChaCha when used as a stream cipher. It is in theory possible
-    /// to use `set_counter` to obtain the conventional ChaCha pseudorandom
-    /// stream associated with a particular nonce. This is not a supported use
-    /// of the RNG, because a nonce set that way is not treated as a constant
-    /// value but still as part of the counter, besides endian issues.
+    /// Get the offset from the start of the stream, in bytes.
     ///
-    /// # Examples
-    ///
-    /// ```rust
-    /// use rand::{ChaChaRng, RngCore, SeedableRng};
-    ///
-    /// // Note: Use `FromEntropy` or `ChaChaRng::from_rng()` outside of testing.
-    /// let mut rng1 = ChaChaRng::from_seed([0; 32]);
-    /// let mut rng2 = rng1.clone();
-    ///
-    /// // Skip to round 20. Because every round generates 16 `u32` values, this
-    /// // actually means skipping 320 values.
-    /// for _ in 0..(20*16) { rng1.next_u32(); }
-    /// rng2.set_counter(20, 0);
-    /// assert_eq!(rng1.next_u32(), rng2.next_u32());
-    /// ```
-    pub fn set_counter(&mut self, counter_low: u64, counter_high: u64) {
-        self.0.inner_mut().set_counter(counter_low, counter_high);
-        self.0.reset(); // force recomputation on next use
+    /// This returns an error only if the result is not representable in the
+    /// 128-bit return value.
+    #[cfg(feature = "i128_support")]
+    pub fn get_offset(&self) -> u128 {
+        let core = self.0.inner();
+        let mut c = (core.state[13] as u128) << 32
+                  | (core.state[12] as u128);
+        let mut index = self.0.index();
+        // c is the end of the last block generated, unless index is at end
+        if index >= STATE_WORDS {
+            index = 0;
+        } else {
+            c -= 1;
+        }
+        // buffer is 16 words each of 4 bytes:
+        (c << 6) | (index << 2) as u128
     }
 
-    /// Sets the number of rounds to run the ChaCha core algorithm per block to
-    /// generate.
-    ///
-    /// By default this is set to 20. Other recommended values are 12 and 8,
-    /// which trade security for performance. `rounds` only supports values
-    /// that are multiples of 4 and less than or equal to 20.
-    ///
-    /// # Examples
-    ///
-    /// ```rust
-    /// use rand::{ChaChaRng, RngCore, SeedableRng};
-    ///
-    /// // Note: Use `FromEntropy` or `ChaChaRng::from_rng()` outside of testing.
-    /// let mut rng = ChaChaRng::from_seed([0; 32]);
-    /// rng.set_rounds(8);
-    ///
-    /// assert_eq!(rng.next_u32(), 0x2fef003e);
-    /// ```
-    pub fn set_rounds(&mut self, rounds: usize) {
-        self.0.inner_mut().set_rounds(rounds);
-        self.0.reset(); // force recomputation on next use
+    /// Set the offset from the start of the stream, in bytes.
+    #[cfg(feature = "i128_support")]
+    pub fn set_offset(&mut self, offset: u128) -> Result<(), ()> {
+        if (offset & 0x3) != 0 || (128 - offset.leading_zeros() - 6 > 64) {
```

Comment:
> `(128 - offset.leading_zeros() - 6 > 64)`

I think it is ok to just wrap around if the offset is beyond the period. I.e. just mask out the first couple of bits?

Can we define the offset/position to be in multiples of the word size? Users have to care about it already, so we may as well make the function infallible. I think there is just about no RNG that can seek to an exact byte, always to something word-size.

Reply:
True, I suppose we wrap anyway during usage!

Seeking to word size also makes a lot of sense given that we're trying to "standardise" this such that `fill_bytes` consumes a whole number of words.